mongodb-plugin/src/main/java/io/cdap/plugin/BSONConverter.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin;
+
+import com.google.common.collect.Lists;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.bson.BSONObject;
+import org.bson.types.BasicBSONList;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Converts {@link BSONObject} to {@link StructuredRecord}.
+ */
+public class BSONConverter {
+  private static final List<Schema.Type> VALID_TYPES = Lists.newArrayList(Schema.Type.ARRAY, Schema.Type.BOOLEAN,
+                                                                          Schema.Type.BYTES, Schema.Type.STRING,
+                                                                          Schema.Type.DOUBLE,
+                                                                          Schema.Type.FLOAT, Schema.Type.INT,
+                                                                          Schema.Type.LONG, Schema.Type.NULL);
+  private final Schema schema;
+
+  public BSONConverter(Schema schema) throws IOException {
+    this.schema = schema;
+  }
+
+  public StructuredRecord transform(BSONObject bsonObject) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    for (Schema.Field field : schema.getFields()) {
+      builder.set(field.getName(), extractValue(bsonObject.get(field.getName()), field.getSchema()));
+    }
+    return builder.build();
+  }
+
+  public static void validateSchema(Schema schema) {
```

Comment:
this class should not be responsible for validation

Reply:
Moved to [MongoDBSourceConfig#validate](https://github.com/vladhlinskiy/database-plugins/pull/14/files#diff-56e5c8c248fc520b3c862df33668d2feR187).